README.md

![epispot](https://i.ibb.co/hXMjrCV/epi-spot.png)

---

# epispot v3

A Python package for the mathematical modeling of infectious diseases via  compartmental models. Originally designed for epidemiologists, epispot can be adapted for almost any type of modeling scenario.

## Features

The epispot package currently only supports compartmental models, though we plan to expand the package to work for stochastic agent-based and spatial models as well. Currently, epispot offers the following:

- Quick compilation of compartmental models with the following compartments:
    - Susceptible
    - Infected
    - Recovered
    - Removed
    - Exposed
    - Dead
    - Critical
    - Hospitalized
- Custom-defined compartments for research
- Built-in graphing and visualization engine
    - Plots model predictions interactively
    - Creates comparisons between models

Due to its diverse range of features, epispot can be used for both research and experimental modeling. If you would like to add more modeling support, please see the [contributing section](#contributing).

## Installation

**Notice**: The epispot nightly package is now officially deprecated.

The epispot package can be installed from PyPI, Anaconda, or be built from the source. Below are the steps for each installation process:

### PyPI

This is the easiest way to install epispot. From a terminal, enter:

```shell
pip install epispot
```

### Anaconda

The standard version of epispot is published to `conda` using the `conda-forge` channel. To install, please use:

```shell
conda config --add channels conda-forge
conda install -c conda-forge epispot
```

### Building from the source

This is the hardest way to install epispot and it is recommended that you use either PyPI or Anaconda to install it instead. However, if you would like to contribute to the repository, this will be particularly useful.

Clone the repository and install the dependencies with:

```shell
git clone https://github.com/epispot/epispot
cd epispot
pip install -r requirements.txt
pip install -r bin/requirements.txt
```

Then, build the latest version with:

```shell
python setup.py install
```

For older releases, first checkout a version tag before installing; e.g.:

```shell
git checkout v2.1.1
python setup.py install
```

## Quick Demo

Using epispot in a Python REPL to create the well-known SIR model (in less than three lines of code):

![Quick demo GIF](https://epispot.github.io/epispot/assets/demo.gif)

## Documentation

Right now, documentation for some of the latest releases is quite shaky.
The official docs are located at <https://epispot.github.io/epispot/> but may be incomplete in many areas.

## Examples

The GitHub repository has a vast array of samples using epispot. You can start by checking out the [`explorables/`](./explorables/) directory. In it, you'll find many programs designed to help you get started with epispot and various hands-on examples.

## Badges

**PyPI**: ![latest-release](https://shields.mitmproxy.org/pypi/v/epispot.svg?color=success)
[![Downloads](https://pepy.tech/badge/epispot)](https://pepy.tech/project/epispot)  
**LGTM**: [![Language grade: Python](https://img.shields.io/lgtm/grade/python/g/quantum9Innovation/epispot.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/quantum9Innovation/epispot/context:python)
[![Total alerts](https://img.shields.io/lgtm/alerts/g/quantum9Innovation/epispot.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/quantum9Innovation/epispot/alerts/)  
**Codecov**: [![codecov](https://codecov.io/gh/epispot/epispot/branch/master/graph/badge.svg?token=WGIM127RFY)](https://codecov.io/gh/epispot/epispot)

## Feedback

If you have any feedback, please

- Create a discussion on GitHub
- Create an issue if you've found a bug
- Submit a PR if you want to add a new feature
- Contact a [CODEOWNER](https://github.com/epispot/epispot/tree/master/.github/CODEOWNERS)

## Contributing

Contributions are always welcome!
See [CONTRIBUTING.md](https://github.com/epispot/epispot/tree/master/CONTRIBUTING.md) for instructions on how to get started, including environment setup and instructions to build from the source. Please note also that epispot has many guides dedicated to certain types of
contributions. Please see

- [DOCUMENTATION.md](https://github.com/epispot/epispot/tree/master/DOCUMENTATION.md) for documentation additions
- [SECURITY.md](https://github.com/epispot/epispot/tree/master/SECURITY.md) for epispot's security policy

## Citation

[![GPLv3 License](https://img.shields.io/badge/License-GPL%20v3-green.svg)](https://opensource.org/licenses/)

If you plan on using epispot in your project, please abide by the GPLv3 license. This requires that any changes you make to epispot are open-sourced under the GPLv3 license as well and that you give credit to the author, which you can do by citing the project in your research, linking back to the original repository, or mentioning the author @quantum9innovation.

For research, you can also use epispot's DOI to reference the project:
> [![DOI](https://zenodo.org/badge/280527664.svg)](https://zenodo.org/badge/latestdoi/280527664)

The recommended citation for epispot is:
> quantum9innovation (2021, April 2). epispot/epispot: (Version 2.1.1).
> Zenodo. <http://doi.org/10.5281/zenodo.4624423>

## Related Work

There are many related projects to epispot, although we believe that epispot has greater extensibility than many of these other projects. Additionally, epispot is quite portable (graphs can be used as web components, displayed as images, etc.). However, below we would like to acknowledge a few projects that may be better suited to certain use cases:

- [EpiJS](https://github.com/Quantalabs/EpiJS) by @Quantalabs contains the same functionality of epispot but in pure JavaScript, making it more suitable for use in a web browser
- [covasim](https://github.com/institutefordiseasemodeling/covasim) by the [Institute for Disease Modeling](https://www.idmod.org) offers agent-based stochastic models as opposed to epispot's compartmental modeling structure
- [CovsirPhy](https://github.com/lisphilar/covid19-sir) by @lisphilar offers greater support for loading and analyzing real COVID-19 data instead of running simulations

## Authors

Please see our [CODEOWNERS](https://github.com/epispot/epispot/tree/master/.github/CODEOWNERS) file for authors. Because epispot is an open-source project, different pieces of our code have different authors. However, if citing epispot or using it in another project, you can cite @quantum9innovation as the lead author.

## Acknowledgements

### Idea & Inspiration

The original idea for epispot came from a [3Blue1Brown video](https://www.youtube.com/watch?v=gxAaO2rsdIs) on basic infectious disease dynamics and an [interactive article](https://www.washingtonpost.com/graphics/2020/world/corona-simulator/) in the Washington Post. This in turn inspired the very basic infectious disease dynamics simulated [here](https://quantum9innovation.github.io/disease/). However, what finally set the package into motion was a series of articles by Henry Froese, available on Medium
[here](https://towardsdatascience.com/infectious-disease-modelling-part-i-understanding-sir-28d60e29fdfc),
along with their corresponding interactive notebooks.

### Code Development & Maintenance

The epispot project is built on open source code and is itself open-source. The initial core development was fueled by @quantum9innovation and much of the codebase was maintained by @Quantalabs. Additionally, thank you to all of epispot's open-source contributors!

![epispot's open-source contributors](https://contrib.rocks/image?repo=epispot/epispot)


CHANGELOG.md
# Release Notes

Latest stable release: 2.1.1\
Latest nightly release: 2.1.1.15\
Latest alpha release: 3.0.0-alpha-3

Releases are listed from most recent to least recent. All alpha and in-development versions are released to the [epispot-nightly](https://pypi.org/project/epispot-nightly/) project while all other releases are shipped to the [main project](https://pypi.org/project/epispot/).

---

## Version Support

Below is the official list of epispot versions and their support status. If we plan on deprecating a version, there will be a note next to the version listing a date (or approximate date) for when the version will be officially unsupported. The current development version is in bold.

| Version | Patch | Bugfix | Security | Notes |
| ------- | ----- | ------ | -------- | ----- |
| 2.1   | 2.1.1 | ✔️ | ✔️ | Latest stable release |
| 2.0   | 2.0.2 | :x: | ✔️ |
| <= 1.1   | 1.1.0 | :x: | :x: | Deprecated |
| nightly latest   | 2.1.1.15 | :x: | :x: | Deprecated |
| **3.0.0-alpha** | **3.0.0a3** | **✔️** | **✔️** | **Supported until beta release of v3** |
| nightly < latest  | 2.1.1.x | :x: | :x: | Deprecated |

---

## Release Schedule

For epispot v3.0.0:

| Release | Deadline | End of life |
| ------- | ------- | ----------- |
| v3.0.0-alpha-1 | N/A | 12/31/21 |
| v3.0.0-alpha-2 | N/A | 12/31/21 |
| v3.0.0-alpha-3 | 10/18/21 | 12/31/21 |
| ~~v3.0.0-beta~~ | skipped | N/A |
| v3.0.0 | **pending** | LTS |

---

## 3.0.0-alpha-3

As the third release in the v3-alpha series, this version updates all dependencies and brings Python 3.10 support to epispot.

Epispot will now support all Python versions from 3.7 to 3.10, with our focus being on Python 3.10. [All dependencies have also  been updated](https://github.com/epispot/epispot/pull/113) to ensure compatibility with Python 3.10 and to catch up on missed dependency updates from last release.

Additionally, in minor updates, epispot will be removing unnecessary CI checks (e.g. [DeepSource](https://github.com/epispot/epispot/commit/6119238737e088aceecb99c9bcfc57644f5c322f)) from analyzing non-package files and adding more [detailed summaries for Zenodo uploads](https://github.com/epispot/epispot/commit/95c99a82050e3ed20bdd73567e1d36b6c38bc766/).

Finally, with this release comes a major update to epispot's security policy, which you can view [here](https://github.com/epispot/epispot/blob/master/SECURITY.md). This means that our entire organization will now follow the [Google OSS Vulnerability Guide](https://github.com/google/oss-vulnerability-guide), which will be applied for any security vulnerabilities we find in any epispot versions.

## 3.0.0-alpha-2 (standard-params)

This is the second stage of the v3 release rollout, following [v3.0.0-alpha-1](#300-alpha-1-massive-plots). The major change in this release revolves around solving issue [#73](https://github.com/epispot/epispot/issues/73), which completely redesigns epispot's internals.

In deprecations, we are now officially deprecating the entire `epispot.fitters` module. These changes have been reflected in our [updated documentation](https://epispot.github.io/epispot/en/v3.0.0-alpha-2/fitters.html). New alternatives are expected to arrive soon. Additionally, we've made the decision to officially end support for Python 3.6, following its removal from our testing suite and manifest files (including, finally, `setup.cfg`).

The big news here, of course, are the huge changes to epispot's `Model` and `Compartment` objects. In short, `epispot.models` and `epispot.comps` have been completely redesigned so that each compartment in `epispot.comps` is a sub-compartment of a `Compartment` object, allowing you to create your own custom compartments. Additionally, models now accept the parameters and pass them on to compartments, instead of the other way around. This change is huge and is a large deprecation, so there is a lot to cover. We strongly recommend you read more about this change [here](https://github.com/epispot/epispot/issues/73) or read the new docs.

In minor updates, we've [added support for Dependabot](https://github.com/epispot/epispot/issues/79) and [bumped Numpy to 1.21.1](https://github.com/epispot/epispot/commit/2fb5eff59c3b9d77f22b6dd1f95d34a9ac1bce6c#diff-9a3d09936710783b0cc2e50f54f8cc456be41c432647337fcf9a9391a9e81b98), featuring small performance improvements on the 1.21.0 release. In fact, there's now a `pre-requirements.txt` file to aid manual installation of the epispot package. Additionally, we've [added spellcheck as a routine CI/CD procedure](https://github.com/epispot/epispot/pull/92) on epispot and ensured that our new versions are typo-free. Speaking of CI/CD, we've officially [made the switch *back* from Travis CI to GitHub Actions](https://github.com/epispot/epispot/pull/93) which will give us faster build times and greater reliability. After the first v3 alpha release, we have now also [included alpha releases in our security policy](https://github.com/epispot/epispot/commit/43449d362eab94444a808fb6cedf6f04caee6cf0), so be sure to [check out the new security policy](https://github.com/epispot/epispot/blob/master/SECURITY.md).

Finally, epispot now comes with a
[new built-in sanity check](https://github.com/epispot/epispot/commit/71a70a040b8c60a77e038eb1edee0dda785798ef#diff-3bd065e1fc4a45ad5e94ee148eaf369a81d39db0c3ad4847b0d7323e4fe16a71),
*but it won't run automatically*. This will hopefully increase performance by just a little bit since no checks are being run (unless they are manually invoked).

## 3.0.0-alpha-1 (massive-plots)

This is the first stage of the v3 release rollout. Following this, there will likely be many more `alpha-x` releases. The major changes in this release come from an epispot operation known as `massive-plots`.


epispot/estimates/data.py
"""
The `epispot.data` module contains all of the estimates pulled from the
literature regarding the parameters of specific diseases. They are 
classified into a hierarchy of classes:

1. `epispot.estimates.utils.Disease` represents a specific disease (e.g. COVID-19)
2. `epispot.estimates.utils.Paper` represents a specific study published in the 
   literature regarding a specific disease (e.g. Ganyani, et al. 2020)
3. `epispot.estimates.utils.Estimate` represents a specific parameter estimate
    presented in a particular paper about a specific disease 
    (e.g. the recovery rate of COVID-19)

"""

# imports
from . import utils
from . import dt


# COVID-19

# Estimates
beta_santos = utils.Estimate(
    id='beta',
    dist=lambda t: 2.93,
    name='Constant-valued approximation of Beta',
    description='Estimated using an SEIR model and fit via a genetic algorithm using data from the Philippines',
)
delta_santos = utils.Estimate(
    id='delta',
    dist=lambda t: 0.28,
    name='Constant-valued approximation of Delta',
    description='Estimated using an SEIR model and fit via a genetic algorithm using data from the Philippines; referred to in the paper as "sigma," though it still represents the incubation rate delta',
)
gamma_santos = utils.Estimate(
    id='gamma',
    dist=lambda t: 0.33,
    name='Constant-valued approximation of Gamma',
    description='Estimated using an SEIR model and fit via a genetic algorithm using data from the Philippines; represents the removal rate in particular (not recovery rate)',
)
kappa_tsay = utils.Estimate(
    id='kappa',
    dist=lambda t: 0.2,
    name='Average testing rate',
    description='Estimated using data from the U.S. in the context of an SEAIR model; represents the probability of asymptomatic individuals becoming officially confirmed as infected cases',
)
gamma_tsay = utils.Estimate(
    id='gamma',
    dist=lambda t: 0.0255,
    name='Average recovery rate',
    description='Estimated using data from the U.S. in the context of an SEAIR model; referred to as "beta" in the paper',
)
rho_tsay = utils.Estimate(
    id='rho',
    dist=lambda t: 0.0255,
    name='Average death rate',
    description='Estimated using data from the U.S. in the context of an SEAIR model; referred to as "mu" in the paper',
)
r0_bentout = utils.Estimate(
    id='R_0',
    dist=lambda t: 4.1,
    name='Estimated initial R Naught',
    description='Estimated using data from the beginning of Algeria\'s COVID-19 outbreak in the context of an SEIR model',
)
beta_bentout = utils.Estimate(
    id='beta',
    dist=lambda t: 0.41,
    name='Estimated initial Beta',
    description='Estimated using data from the beginning of Algeria\'s COVID-19 outbreak in the context of an SEIR model',
)
gamma_bentout = utils.Estimate(
    id='gamma',
    dist=lambda t: 0.1,
    name='Gamma approximation (pulled from various sources)',
    description='Estimated using data from the beginning of Algeria\'s COVID-19 outbreak in the context of an SEIR model; represents the removal (not recovery) rate',
)
delta_bentout = utils.Estimate(
    id='delta',
    dist=lambda t: 0.2,
    name='Delta approximation (pulled from various sources including the WHO Coronavirus dataset)',
    description='Estimated using data from the beginning of Algeria\'s COVID-19 outbreak in the context of an SEIR model; referred to as "lambda" in the paper',
)
gamma_mehra = utils.Estimate(
    id='gamma',
    dist=lambda t: 0.222,
    name='Average Gamma estimate',
    description='Estimated using data from South Korea and the U.S. in the context of a SQAIR model; represents the recovery rate; referred to as "g" in the paper',
)
rho_mehra = utils.Estimate(
    id='rho',
    dist=lambda t: 0.0257,
    name='Average Rho estimate',
    description='Estimated using data from South Korea and the U.S. in the context of a SQAIR model; represents the death rate; referred to as "mu sub d" in the paper',
)
kappa_mehra = utils.Estimate(
    id='kappa',
    dist=lambda t: 0.214,
    name='Average testing rate estimate',
    description='Estimated using data from South Korea and the U.S. in the context of a SQAIR model; represents the probability of asymptomatic individuals becoming officially confirmed as infected cases; referred to as "alpha" in the paper but renamed to kappa for consistency',
)

# Papers
santos = utils.Paper(
    ('Santos', 2022),
    params=[
        beta_santos,
        delta_santos,
        gamma_santos,
    ],
    metadata={
        'title': 'Parameter Estimation for a Modified SEIR Model of the COVID-19 Dynamics in the Philippines using Genetic Algorithm',
        'authors': ('Gabriel Lorenzo I. Santos'),
        'journal': 'medRxiv',
        'date': dt.datetime(2022, 5, 19),
        'url': 'https://www.medrxiv.org/content/10.1101/2022.05.17.22275187v1',
    },
    full='Santos, G. L. I. (2022). Parameter estimation for a modified SEIR model of the COVID-19 dynamics in the Philippines using genetic algorithm (p. 2022.05.17.22275187). medRxiv. https://www.medrxiv.org/content/10.1101/2022.05.17.22275187v1',
)
tsay = utils.Paper(
    ('Tsay', 'et al.', 2020),
    params=[
        kappa_tsay,
        gamma_tsay,
        rho_tsay,
    ],
    metadata={
        'title': 'Modeling, state estimation, and optimal control for the US COVID‑19 outbreak',
        'authors': ('Calvin Tsay', 'Fernando Lejarza', 'Mark A. Stadtherr', 'Michael Baldea'),
        'journal': 'Nature',
        'date': dt.datetime(2020, 7, 1),
        'url': 'https://www.fmda.org/COVID/Published-Peer-Reviewed-Journal-Articles/s41598-020-67459-8.pdf',
    },
    full='Tsay, C., Lejarza, F., Stadtherr, M. A., & Baldea, M. (2020). Modeling, state estimation, and optimal control for the US COVID-19 outbreak. Scientific Reports, 10(1), 10711. https://doi.org/10.1038/s41598-020-67459-8',
)      
bentout = utils.Paper(
    ('Bentout', 'et al.', 2020),
    params=[
        r0_bentout,
        beta_bentout,
        gamma_bentout,
        delta_bentout,
    ],
    metadata={
        'title': 'Parameter estimation and prediction for coronavirus disease outbreak 2019 (COVID-19) in Algeria',
        'authors': ('Soufiane Bentout', 'Abdennasser Chekroun', 'Toshikazu Kuniya'),
        'journal': 'AIMS Public Health',
        'date': dt.datetime(2020, 5, 22),
        'url': 'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7327392/',
    },
    full='Bentout, Soufiane, Abdennasser Chekroun, and Toshikazu Kuniya. “Parameter Estimation and Prediction for Coronavirus Disease Outbreak 2019 (COVID-19) in Algeria.” AIMS Public Health 7, no. 2 (May 22, 2020): 306–18. https://doi.org/10.3934/publichealth.2020026.',
)
mehra = utils.Paper(
    ('Mehra', 'et al.', 2020),
    params=[
        gamma_mehra,
        rho_mehra,
        kappa_mehra,
    ],
    metadata={
        'title': 'Parameter Estimation and Prediction of COVID-19 Epidemic Turning Point and Ending Time of a Case Study on SIR/SQAIR Epidemic Models',
        'authors': ('Amir Hossein Amiri Mehra', 'Mohsen Shafieirad', 'Zohreh Abbasi', 'Iman Zamani'),
        'journal': 'Hindawi',
        'date': dt.datetime(2020, 12, 29),
        'url': 'https://www.hindawi.com/journals/cmmm/2020/1465923/',
    },
    full='Amiri Mehra, Amir Hossein, Mohsen Shafieirad, Zohreh Abbasi, and Iman Zamani. “Parameter Estimation and Prediction of COVID-19 Epidemic Turning Point and Ending Time of a Case Study on SIR/SQAIR Epidemic Models.” Computational and Mathematical Methods in Medicine 2020 (December 29, 2020): e1465923. https://doi.org/10.1155/2020/1465923.',
)

# Initialize
Covid = utils.Disease(
    id='SARS-CoV-2',
    papers=[
        santos,
        tsay,
        bentout,
        mehra,
    ],
    name='COVID-19',
    description='severe acute respiratory syndrome coronavirus 2',
)
